### PR TITLE
Opencl instrumentation

### PR DESCRIPTION
**Pull request by Jokeren**
Code review

### CODE REVIEW

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 1]
```diff
@@ -181,6 +202,96 @@
 // opencl function pointers for late binding
 //----------------------------------------------------------
 
+OPENCL_FN
+(
+  clBuildProgram, 
+  (
+	 cl_program program,
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:209]: indentation

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 2]
```diff
@@ -216,15 +327,140 @@ OPENCL_FN
 );
 
 
-
 static atomic_ullong opencl_pending_operations;
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:330]: Better to have initializers

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 3]
```diff
@@ -216,15 +327,140 @@ OPENCL_FN
...
+
+// we are dumping the debuginfo since the binary does not have debugsection
+static void
+clBuildProgramCallback
+(
+	cl_program program,
+	void* user_data
+)
+{
+#if 0
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:413]: Remove this function

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 4]
```diff
@@ -461,6 +698,222 @@ opencl_bind
 }
 
 
+cl_program
+clCreateProgramWithSource
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:702]: Remove this function

File: src/tool/hpcrun/gpu/opencl/opencl-api.h [hunk 5]
```diff
@@ -61,6 +61,13 @@
 // interface operations
 //******************************************************************************
 
+char*
+getDebugInfoFullFileName
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-api.h:65]: Remove this function

File: src/tool/hpcrun/gpu/opencl/opencl-intercept.c [hunk 6]
```diff
@@ -47,317 +47,43 @@
...
 (
-  void
+ void
 )
 {
 #ifndef HPCRUN_STATIC_LINK
   ETMSG(OPENCL, "setting up opencl intercepts");
-  gotcha_wrap(opencl_bindings, 4, "opencl_bindings");
-  opencl_intercept_initialize();
+	gpu_metrics_KER_BLKINFO_enable();
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-intercept.c:77]: indentation

File: src/tool/hpcrun/gpu/opencl/opencl-intercept.h [hunk 7]
```diff
@@ -113,6 +113,27 @@ typedef cl_int (*clwritebuffer_t)(
 );
 
 
+typedef cl_int (*clbuildprogram_t)
```
Comment (Jokeren) [inline src/tool/hpcrun/gpu/opencl/opencl-intercept.h:116]: Shall we keep these two types?

**Review by Jokeren — COMMENTED** [review]
I will address these issues

File: src/lib/banal/Struct-Output.hpp [hunk 8]
```diff
@@ -78,6 +78,43 @@ void printFileEnd(ostream *, FileInfo *);
 void printProc(ostream *, ostream *, string, FileInfo *, GroupInfo *,
 	       ProcInfo *, HPC::StringTable & strTab);
 
+void printBlockAndInstructionOffset(ostream * os, string file_name);
+
```
Comment (jmellorcrummey) [inline src/lib/banal/Struct-Output.hpp:82]: these two functions don't belong in a header file.

moreover, they are general utilities. they don't belong in Struct-Output.cpp either. who uses them?

Comment (Jokeren) [reply src/lib/banal/Struct-Output.hpp:82]: I've already marked them. I'll remove them in the upcoming update.

**Comment by blue42u** [comment]
@mxz297 Please double-check that OpenCL and Level0 are still working correctly with this merge, I think I caught everything but I don't have the tests for it.

Also compiling with Dyninst 10.2.1 fails, I believe it failed before the merge as well:
```console
.../src/lib/banal/gpu/GPUBlock.cpp: In member function ‘virtual void Dyninst::ParseAPI::GPUBlock::getInsns(Dyninst::ParseAPI::Block::Insns&) const’:
.../src/lib/banal/gpu/GPUBlock.cpp:24:24: error: ‘intel_gpu_op_general’ was not declared in this scope
   24 |     entryID entry_id = intel_gpu_op_general;
      |                        ^~~~~~~~~~~~~~~~~~~~
```

**Comment by mxz297** [comment]
@blue42u I will test things along the way. 

Yes, we will require dyninst master until Dyninst 11.0 is released.

**Pull request merged**